lib/code_corps/github.ex

```diff
@@ -1,41 +1,245 @@
+  @base_access_token_params %{
+    client_id: @client_id,
+    client_secret: @client_secret
+  }
 
-  Returns one of the following:
+  defmodule APIErrorObject do
+    @moduledoc """
+    Represents an error object from the API.
 
-  - `{:ok, %CodeCorps.User{}}`
-  - `{:error, %Ecto.Changeset{}}`
-  - `{:error, "some_github_error"}`
-  """
-  @spec connect(User.t, String.t, module) :: {:ok, User.t} | {:error, String.t}
-  def connect(%User{} = user, code, api \\ @api) do
-    case code |> api.connect do
-      {:ok, github_auth_token} -> user |> associate(%{github_auth_token: github_auth_token})
-      {:error, error} -> {:error, error}
+    See the type documentation for more about how this struct can be
+    used.
+
+    The full details of error objects can be found in the
+    [GitHub API documentation](https://developer.github.com/v3/#client-errors).
+    """
+    defstruct [:code, :field, :resource]
+
+    def new(opts) do
+      struct(__MODULE__, opts)
+    end
+  end
+
+  defmodule APIError do
+    defstruct [:documentation_url, :errors, :message, :status_code]
+
+    @type t :: %__MODULE__{
+      documentation_url: String.t | nil,
+      errors: List.t | nil,
+      message: String.t | nil,
+      status_code: pos_integer | nil
+    }
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message, "errors" => errors}}) do
+      errors = Enum.into(errors, [], fn error -> convert_error(error) end)
+
+      %__MODULE__{
+        errors: errors,
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message, "documentation_url" => documentation_url}}) do
+      %__MODULE__{
+        documentation_url: documentation_url,
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message}}) do
+      %__MODULE__{
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    defp convert_error(%{"code" => code, "field" => field, "resource" => resource}) do
+      APIErrorObject.new([code: code, field: field, resource: resource])
+    end
+  end
+
+  defmodule HTTPClientFailed do
+    defstruct [:reason, message: """
+    The GitHub HTTP client encountered an error while communicating with
+    the GitHub API.
+    """]
+
+    def new(opts) do
+      struct(__MODULE__, opts)
+    end
+  end
+
+  @type method :: :get | :post | :put | :delete | :patch
+  @type headers :: %{String.t => String.t} | %{}
+  @type body :: {:multipart, list} | map
+  @typep http_success :: {:ok, integer, [{String.t, String.t}], String.t}
+  @typep http_failure :: {:error, term}
+
+  @type api_error_struct ::
+    %APIError{}            |
+    %HTTPClientFailed{}
+
+  @spec get_base_url() :: String.t
+  defp get_base_url() do
+    Application.get_env(:code_corps, :github_base_url) || "https://api.github.com/"
+  end
```

Comment:
Best way I could think to allow us to inject a different Bypass server endpoint, borrowed from an article on Bypass.